[JENKINS-70301] Do not report implied dependencies for WMI Windows Agents plugin

MarkEWaite opened the pull request:
## Do not report implied dependencies for WMI Windows Agents plugin

[JENKINS-70301](https://issues.jenkins.io/browse/JENKINS-70301) notes that the [WMI Windows Agents plugin](https://plugins.jenkins.io/windows-slaves/) (id:windows-slaves) was split from core in 1.547 and is now deprecated.  Because it is deprecated, users want to remove it.  They cannot remove it if they have plugins installed that depend on a Jenkins version prior to 1.547.

Only two plugins in the jenkinsci GitHub organization actually depend on any classes from the WMI Windows Agent plugin.

* [cli-extras-plugin](https://github.com/jenkinsci/cli-extras-plugin) - never released, no tags, no install history
* [multi-slave-config](https://github.com/jenkinsci/multi-slave-config-plugin) - already requires windows-slaves plugin explicitly and already requires a Jenkins core version newer than 1.547. 1340 installations, last released 8 years ago

Rather than require updates of all the plugins that require a Jenkins version less than 1.547, this removes the windows-slaves plugin from the split plugins list.  Users will not be warned that their plugin might have a dependency on windows-slaves because the analysis has confirmed that there is only 1 delivered plugin with that dependency, and it is an explicitly declared dependency with a newer Jenkins minimum version than the version from which the WMI Windows Agents plugin was split from Jenkins core.

[A google sheet](https://docs.google.com/spreadsheets/d/1KDIJIu41rIRlj6PjzyhAY9vazKmd3PVwqM3_-32Lt-o/edit?usp=sharing) lists the plugins that have an implied dependency on the WMI Windows Agents plugin.

No new automated tests because this is a one-time operation that has been tested interactively.  Plugin compatibility test evaluation:

* https://github.com/jenkinsci/bom/pull/1674 passed

### Testing done

* Installed suggested plugins from wizard and confirmed WMI Windows Agents plugin is not installed
* Installed slave-setup plugin (has implied WMI Windows Agents plugin dependency) and confirmed that WMI Windows Agents plugin is not installed
* Installed multi-slave-setup plugin (has explicit WMI Windows Agents plugin dependency) and confirmed that WMI Windows Agents plugin was installed as a dependency
* Remove multi-slave-setup plugin and WMI Windows Agents plugin and confirmed that Jenkins reported no errors or issues
* Installed five plugins with old base versions and confirmed that WMI Windows Agents plugin was not installed.  The five plugins included backup-plugin, cloudbees-credentials-plugin, unity3d-plugin, groovy-label-assignment-plugin, and maven-deployment-linker plugin.
* Performed a backup with the backup plugin and confirmed that it was well behaved
* Removed the windows-slaves plugin as a bundled plugin so that it is not included in the war file.  Confirmed the result war file loaded the five plugins tested earlier and was well behaved before and after those five plugins were uninstalled

### Proposed changelog entries

- Do not report implied dependencies for WMI Windows Agents plugin.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [x] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [x] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [x] For new APIs and extension points, there is a link to at least one consumer.
- [x] Run plugin compatibility tester with an incremental build

### Desired reviewers

@jglick or @bcrow

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [x] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [x] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [x] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [x] Proper changelog labels are set so that the changelog can be generated automatically.
- [x] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [x] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).


<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7568"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

MarkEWaite committed:
Hide split of windows-slaves from other plugins

https://issues.jenkins.io/browse/JENKINS-70301 notes that the WMI Windows
Agents plugin (id:windows-slaves) was split from core in 1.547 and is
now deprecated.  Because it is deprecated, users want to remove it.
They cannot remove it if they have plugins installed that depend on a
Jenkins version prior to 1.547.

Only two plugins actually depend of any classes from the WMI Windows
Agent plugin.

* cli-extras-plugin - never released, no tags, no install history
* multi-slave-config - already requires windows-slaves plugin explicitly
  and already requires a Jenkins core version newer than 1.547. 1340
  installations, last released 8 years ago

Rather than require updates of all the plugins that require a Jenkins
version less than 1.547, this removes the windows-slaves plugin from the
split plugins list.  Users will not be warned that their plugin might
have a dependency on windows-slaves because the analysis has confirmed
that there is only 1 delivered plugin with that dependency, and it is
an explicitly declared dependency with a newer Jenkins minimum version
than the version from which the WMI Windows Agents plugion was split.

Testing done

* Installed suggested plugins from wizard and confirmed WMI Windows
  Agents plugin is not installed

* Installed slave-setup plugin (has implied WMI Windows Agents plugin
  dependency) and confirmed that WMI Windows Agents plugin is not installed

* Installed multi-slave-setup plugin (has explicit WMI Windows Agents
  plugin dependency) and confirmed that WMI Windows Agents plugin was
  installed as a dependency

* Remove multi-slave-setup plugin and WMI Windows Agents plugin and
  confirmed that Jenkins reported no errors or issues

* Installed five plugins with old base versions and confirmed
  that WMI Windows Agents plugin was not installed.  The five plugins
  included backup-plugin, cloudbees-credentials-plugin, unity3d-plugin,
  groovy-label-assignment-plugin, and maven-deployment-linker plugin.

* Performed a backup with the backup plugin and confirmed that it was
  well behaved

https://docs.google.com/spreadsheets/d/1KDIJIu41rIRlj6PjzyhAY9vazKmd3PVwqM3_-32Lt-o/edit?usp=sharing
lists the plugins that have an implied dependency on the WMI Windows
Agents plugin.

basil reviewed (commented):
This change seems fine, but I am afraid it is not a one-time operation as you write in the PR description. Users have complained about this type of problem in other contexts and with other detached plugins, so a general solution would be desirable. However I will not insist on that for this PR.

I am not yet approving this PR because I think it is missing a change to stop bundling `windows-slaves` in the WAR file in `war/pom.xml`. It does not make sense to bundle a plugin that is not used, so I think we should stop bundling it in the WAR file. Once that fix is made I would be ready to approve this PR.

Since detached plugin shenanigans are a frequent source of PCT failures, I think it would be prudent to run PCT on this before integrating, but I will not insist on it.

MarkEWaite committed:
Remove windows-slaves from war file

No need to bundle a detached plugin that has no consumers of its API
that are based on a Jenkins version before the plugin was split from core.

https://github.com/jenkinsci/jenkins/pull/7568#pullrequestreview-1238137930
detected that the bundling was left in by the original commit.

basil commented on war/pom.xml:
```diff
@@ -305,13 +305,6 @@ THE SOFTWARE.
                   <version>3.1.5</version>
                   <type>hpi</type>
                 </artifactItem>
-                <artifactItem>
-                  <!-- detached after 1.547 -->
-                  <groupId>org.jenkins-ci.plugins</groupId>
-                  <artifactId>windows-slaves</artifactId>
-                  <version>1.8.1</version>
-                  <type>hpi</type>
-                </artifactItem>
```
It is helpful to run `LoadDetachedPluginsTest#noUpdateSiteWarnings` when updating this list, but unfortunately that test has to be `@Ignore`d by default to avoid failing unrelated builds when new security warnings are introduced. As a courtesy I ran that test for you, but it would be helpful to keep this in mind for future changes.

MarkEWaite replied:
Thanks.  As practice for myself, I removed the `@Ignore` on that test method and ran it to confirm that it passes.

NotMyFault reviewed (approved):
/label ready-for-merge
